### ISSUE TITLE
Empty alerts on errors 403

### ISSUE THREAD

**Issue by gmarziou** [jhipster/generator-jhipster#1800]
Hi team,

I had a problem of invalid CSRF cookie, in login screen this resulted in 2 requests in error and 2 empty alerts being displayed. My app was configured with "enableTranslation": true.

POST /api/authentication error 403
{"timestamp":1438467827536,"status":403,"error":"Forbidden","message":"Invalid CSRF Token 'null' was found on the request parameter '_csrf' or header 'X-CSRF-TOKEN'.","path":"/api/authentication"}

POST /api/logout error 403
{"timestamp":1438467827571,"status":403,"error":"Forbidden","message":"Invalid CSRF Token 'null' was found on the request parameter '_csrf' or header 'X-CSRF-TOKEN'.","path":"/api/logout"}

The problem is in alert.directive.js :

~~~
	var addErrorAlert = function (message, key, data) {
		AlertService.error(key, data);
	}
~~~

When entering this function, the parameters were:

~~~
   message = "Invalid CSRF Token 'null' was found on the request parameter '_csrf' or header 'X-CSRF-TOKEN'."
   key = undefined
   data = undefined
~~~

So either, we should display the message or we should not display the alert.

401 and 403 are probably special cases that may require special processing, as the error is correctly handled by the LoginController which displays an error 'Authentication failed...'. 
So here I think no alert should be displayed, but maybe it should be logged.

**Comment by deepu105** [comment]
Oh yes, i guess i was sending the message as fallback for no translation
scenario, but i guess we should put a null check and overwite the message
if key is null, ill do a PR today

**Comment by deepu105** [comment]
@gmarziou can you check if this is fine? im already excluding 401 errors but not sure if we should exclude all 403s can you suggest `!(response.status == 401 && response.data.path.indexOf("/api/account") == 0 )` a similar pattern for 403 if that needs to be excluded.

This is done in errorhandler.interceptor.js

**Comment by gmarziou** [comment]
@deepu105 I tested your patch, yes it fixes the empty message issue but the user experience is not very satisfying because these messages are somehow redundant with the ones from the LoginController.

To reproduce, go to login screen and enter wrong password, you'll get this:

![image](https://cloud.githubusercontent.com/assets/361007/9026318/3838d3de-392b-11e5-8a59-890c64dbb634.png)

I think it's a difficult problem, it's like handling exceptions in Spring MVC, you can have an exception handler defined in your controller and a global exception handler, but when the exception is caught by the controller it does not reach the global handler.

So in this context, maybe there should be an option of the directive to indicate whether the errors should be handled globally or by the controller or maybe in this screen the directive should not be used.

**Comment by deepu105** [comment]
Are you using OAuth?  Ill see how this can be tackled. Which error message do you think is preferred?

**Comment by gmarziou** [comment]
It was either session auth or xauth token but not oauth2.

I think there should not be any notification alert in this screen because the controller already provides the only message that is meaningful to an end-user. 

In fact, I'm not convinced by automatic alerts as implemented by jhAlertToast directive because they are not contextual and because I don't know how to disable them depending on the context.

So to summarize, for me:
- jhAlertToast 's controller should not call addErrorAlert() in default switch statement
- there should be a mechanism to disable jhAlertToast directive globally or per screen

**Comment by deepu105** [comment]
@gmarziou I guess i have to agree about the default in switch part.

1. I didn't think of these as contextual as the idea was to have universal alert which can be used regardless of which view you are in, similar to the concept of push notifications. you can simply add an alert form anywhere and it shows up as a toast. 
2. I guess for now we can remove the default block and do auto alerts only for 400 errors and any other fixed error codes.
3. I guess the remaining notifications in the generated app can be refactored to use AlerService that way the alerts will be consistent 
4. Option to disable alerts globally can be easily done but to do them for a view might be challenging, I dont have any idea for now on that. may be we can create a seperate enhancement issue on that and tackle later

cc @gzsombor

**Comment by renszarv** [comment]
Having a separate 'error-display' directive, which can be added/removed from individual forms could solve the disparity issue on the login form. Of course, that would mean that the nice 'toast' based, global alerts should be disabled.

**Comment by gzsombor** [comment]
Having a separate 'error-display' directive, which can be added/removed from individual forms could solve the disparity issue on the login form. Of course, that would mean that the nice 'toast' based, global alerts should be disabled.

**Comment by deepu105** [comment]
@gzsombor I guess you had it that way originally and I thought that one can be a global addition as it would be easy to maintain.
I think we can have a hybrid,ill move the directive to individual views instead of index.html that way if you do not want it you can just remove the directive. I guess for now ill make it look like existing alerts , if peoplewant to have themas toast they can simply add the css back, ill mention it in the docs

This would solve the concern of Julian as well

what say @gmarziou @gzsombor @jdubois

**Comment by gzsombor** [comment]
:+1:

**Comment by deepu105** [comment]
Hi guys,
done I have split the alert directive into a normal alert `jh-alert` and error alert `jh-alert-error` and have moved these into the entity html files so that these can be removed from a view if not required.

Also changed the look and feel to be consistant with remaining alerts on JH default theme

@gmarziou @jdubois  is this OK now?

![image](https://cloud.githubusercontent.com/assets/1107223/9034015/fe0e8314-39fe-11e5-803e-06db26cb05b3.png)

![image](https://cloud.githubusercontent.com/assets/1107223/9034037/1b420348-39ff-11e5-8a2b-b8dd3164939f.png)

**Comment by gmarziou** [comment]
:+1: 

That's much better, this way we have a good default mechanism but still we give the developper the possibility to do it differently.

**Comment by deepu105** [comment]
@jdubois can i merge this?

**Comment by jdubois** [comment]
@deepu105 oh yes, and this corrects the CSS "alert" class which I didn't like

**Comment by vsarabu** [comment]
Hi, can anyone please share the peace of code how you fixed the issue for red bar with a message.for more information about my problem please follow the link.

http://stackoverflow.com/questions/33202543/custom-error-message-for-bad-request-in-jhipster-project

Thanks in advance
